src/controllers/version/metadata.rs

```diff
@@ -84,3 +104,145 @@ pub async fn show(
+    update_version_yank_state(version, update_data)?;
+    perform_version_yank_update(state, req, conn, version, krate)?;
+
+    Ok(())
+}
+
+fn update_version_yank_state(version: &mut Version, update_data: &VersionUpdate) -> AppResult<()> {
+    match (update_data.yanked, &update_data.yank_message) {
+        (Some(true), Some(message)) => {
+            version.yanked = true;
+            version.yank_message = Some(message.clone());
+        }
+        (Some(yanked), None) => {
+            version.yanked = yanked;
+            version.yank_message = None;
+        }
+        (Some(false), Some(_)) => {
+            return Err(bad_request("Cannot set yank message when unyanking"));
+        }
+        (None, Some(message)) => {
+            if version.yanked {
+                version.yank_message = Some(message.clone());
+            } else {
+                return Err(bad_request(
+                    "Cannot update yank message for a version that is not yanked",
+                ));
+            }
+        }
+        // If both yanked and yank_message are None, do nothing.
+        // This function only cares about updating the yanked state and yank message.
+        (None, None) => {}
+    }
+    Ok(())
+}
+
+pub fn perform_version_yank_update(
+    state: &AppState,
+    req: &Parts,
+    conn: &mut impl Conn,
+    version: &Version,
+    krate: &Crate,
+) -> AppResult<()> {
+    let auth = AuthCheck::default()
+        .with_endpoint_scope(EndpointScope::Yank)
+        .for_crate(&krate.name)
+        .check(req, conn)?;
+
+    state
+        .rate_limiter
+        .check_rate_limit(auth.user_id(), LimitedAction::YankUnyank, conn)?;
+
+    let api_token_id = auth.api_token_id();
+    let user = auth.user();
+    let owners = krate.owners(conn)?;
+
+    if Handle::current().block_on(user.rights(state, &owners))? < Rights::Publish {
+        if user.is_admin {
+            let action = if version.yanked {
+                "yanking"
+            } else {
+                "unyanking"
+            };
+            warn!(
+                "Admin {} is {action} {}@{}",
+                user.gh_login, krate.name, version.num
+            );
+        } else {
+            return Err(custom(
+                StatusCode::FORBIDDEN,
+                "must already be an owner to yank or unyank",
+            ));
+        }
+    }
+
+    // Check if the yanked state or yank message has changed
+    let (yanked, yank_message) = versions::table
+        .find(version.id)
+        .select((versions::yanked, versions::yank_message))
+        .first::<(bool, Option<String>)>(conn)?;
+
+    if yanked == version.yanked && yank_message == version.yank_message {
+        // No changes, return early
+        return Ok(());
+    }
+
+    diesel::update(version)
+        .set((
+            versions::yanked.eq(version.yanked),
+            versions::yank_message.eq(&version.yank_message),
+        ))
+        .execute(conn)?;
```

Comment:
I suggest returning early if there's nothing to change (`(None, None)`) first. Then, make a single update request with a `WHERE` filter that checks if the updating fields are distinct. Finally, examine the returned result to determine if the row was updated.

Reply:
For the patch API, I believe we should consider other updates for the version. So we cannot simply return if the yank update is empty, even if we only have the yank stats update at the moment. I will check GitHub's API design to see what is its behavior if nothing changes.

Reply:
> I see, you are saying if the change is `(None, None)`, then we should skip this query and check at all.

Yeah!

Regarding to sql update part, you should be able to update only when the value is distinct and should also be able to determine if a row was actually updated.